File: hello-go/src/client/core/websocket_client.go
package core

import (
	"encoding/json"
	"fmt"

	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
	"github.com/gorilla/websocket"
)

type WebSocketClient struct {
	conn *websocket.Conn
}

type ControlMessage struct {
	Type     string `json:"type"`
	StreamID string `json:"streamId,omitempty"`
	Offset   *int64 `json:"offset,omitempty"`
	Length   *int   `json:"length,omitempty"`
	Message  string `json:"message,omitempty"`
}

func Connect(uri string) (*WebSocketClient, error) {
	// Configure dialer to disable compression and set larger buffer sizes
	dialer := websocket.Dialer{
		EnableCompression: false,
		WriteBufferSize:   65536,
		ReadBufferSize:    65536,
	}

	conn, _, err := dialer.Dial(uri, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to connect: %w", err)
	}

	return &WebSocketClient{conn: conn}, nil
}

func (c *WebSocketClient) Close() error {
	return c.conn.Close()
}

func (c *WebSocketClient) SendText(message string) error {
	return c.conn.WriteMessage(websocket.TextMessage, []byte(message))
}

func (c *WebSocketClient) SendBinary(data []byte) error {
	return c.conn.WriteMessage(websocket.BinaryMessage, data)
}

func (c *WebSocketClient) ReceiveText() (string, error) {
	msgType, data, err := c.conn.ReadMessage()
	if err != nil {
		return "", fmt.Errorf("failed to receive message: %w", err)
	}
	if msgType != websocket.TextMessage {
		return "", fmt.Errorf("expected text message, got type %d", msgType)
	}
	return string(data), nil
}

func (c *WebSocketClient) ReceiveBinary() ([]byte, error) {
	msgType, data, err := c.conn.ReadMessage()
	if err != nil {
		return nil, fmt.Errorf("failed to receive message: %w", err)
	}
	if msgType == websocket.TextMessage {
		// Log the text message for debugging
		logger.Debug(fmt.Sprintf("Received text message instead of binary: %s", string(data)))
		// This might be an error response, try to parse it
		var msg ControlMessage
		if err := json.Unmarshal(data, &msg); err == nil && msg.Type == "ERROR" {
			return nil, fmt.Errorf("server error: %s", msg.Message)
		}
		return nil, fmt.Errorf("expected binary message, got text: %s", string(data))
	}
	if msgType != websocket.BinaryMessage {
		return nil, fmt.Errorf("expected binary message, got type %d", msgType)
	}
	return data, nil
}

func (c *WebSocketClient) SendControlMessage(msg ControlMessage) error {
	jsonData, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("failed to marshal control message: %w", err)
	}
	logger.Debug(fmt.Sprintf("Sending control message: %s", string(jsonData)))
	return c.SendText(string(jsonData))
}

func (c *WebSocketClient) ReceiveControlMessage() (*ControlMessage, error) {
	text, err := c.ReceiveText()
	if err != nil {
		return nil, err
	}
	logger.Debug(fmt.Sprintf("Received control message: %s", text))

	var msg ControlMessage
	if err := json.Unmarshal([]byte(text), &msg); err != nil {
		return nil, fmt.Errorf("failed to unmarshal control message: %w", err)
	}
	return &msg, nil
}

File: hello-rust/src/client/websocket_client.rs
use anyhow::{Context, Result};
use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use tokio::net::TcpStream;
use tokio_tungstenite::{connect_async, tungstenite::Message, MaybeTlsStream, WebSocketStream};
use tungstenite::{Bytes, Utf8Bytes};

type WsStream = WebSocketStream<MaybeTlsStream<TcpStream>>;

#[derive(Serialize, Deserialize, Debug)]
pub struct ControlMessage {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "streamId")]
    pub stream_id: Option<String>,
    #[serde(rename = "offset")]
    pub offset: Option<u64>,
    #[serde(rename = "length")]
    pub length: Option<usize>,
    #[serde(rename = "message")]
    pub message: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct StreamMessage {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(rename = "streamId")]
    pub stream_id: Option<String>,
    #[serde(rename = "offset")]
    pub offset: Option<u64>,
    #[serde(rename = "length")]
    pub length: Option<usize>,
    #[serde(rename = "message")]
    pub message: Option<String>,
}

pub struct WebSocketClient {
    stream: Option<WsStream>,
}

impl WebSocketClient {
    pub fn new(_uri: &str) -> Self {
        Self {
            stream: None,
        }
    }

    pub async fn connect(&mut self, uri: &str) -> Result<()> {
        let (stream, _) = connect_async(uri)
            .await
            .context(format!("Failed to connect to WebSocket server: {}", uri))?;

        self.stream = Some(stream);
        Ok(())
    }

    pub async fn send_text(&mut self, message: &str) -> Result<()> {
        let stream = self.stream.as_mut().context("Not connected")?;
        stream
            .send(Message::Text(Utf8Bytes::from(message)))
            .await
            .context("Failed to send text message")?;
        Ok(())
    }

    pub async fn send_binary(&mut self, data: Vec<u8>) -> Result<()> {
        let stream = self.stream.as_mut().context("Not connected")?;
        stream
            .send(Message::Binary(Bytes::from(data)))
            .await
            .context("Failed to send binary message")?;
        Ok(())
    }

    pub async fn receive(&mut self) -> Result<Option<Message>> {
        let stream = self.stream.as_mut().context("Not connected")?;
        let msg = stream.next().await;
        match msg {
            Some(result) => Ok(Some(result?)),
            None => Ok(None),
        }
    }

    pub async fn receive_text(&mut self) -> Result<String> {
        let msg = self.receive().await?;

        match msg {
            Some(Message::Text(text)) => Ok(text.to_string()),
            Some(Message::Close(_)) => Ok(String::new()),
            _ => anyhow::bail!("Expected text message, got {:?}", msg),
        }
    }

    pub async fn receive_binary(&mut self) -> Result<Vec<u8>> {
        let msg = self.receive().await?;

        match msg {
            Some(Message::Binary(data)) => Ok(data.to_vec()),
            Some(Message::Close(_)) => Ok(Vec::new()),
            _ => anyhow::bail!("Expected binary message, got {:?}", msg),
        }
    }

    pub async fn close(&mut self) -> Result<()> {
        if let Some(stream) = self.stream.as_mut() {
            stream.close(None)
                .await
                .context("Failed to close WebSocket connection")?;
        }
        Ok(())
    }

    pub async fn send_control_message(&mut self, msg: ControlMessage) -> Result<()> {
        let json = serde_json::to_string(&msg)
            .context("Failed to serialize control message")?;
        self.send_text(&json).await
    }

    pub async fn receive_control_message(&mut self) -> Result<ControlMessage> {
        let text = self.receive_text().await?;
        if text.is_empty() {
            anyhow::bail!("Connection closed");
        }
        serde_json::from_str(&text).context("Failed to parse control message")
    }
}
